### PR TITLE
#1623 Fixed equals() and hashcode()

### PR DESCRIPTION
**Pull request by uditsharma7**
Signed-off-by: Udit Sharma <usharma898@gmail.com>

GitHub issue resolved: #1623 

Briefly describe the changes proposed in this PR:

* Fixed equals and hashcode in BindingSetAssignment
* 
*

### CODE REVIEW

File: core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java [hunk 1]
```diff
@@ -49,13 +49,34 @@
...
 	}
 
 	@Override
 	public int hashCode() {
-		return "BindingSetAssignment".hashCode();
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((bindingNames == null) ? 0 : bindingNames.hashCode());
+		result = prime * result + ((bindingSets == null) ? 0 : bindingSets.hashCode());
+		return result;
```
Comment (hmottestad) [inline core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java]: This looks like a great catch! Any chance that Objects.hash(....) could be used? https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash(java.lang.Object...)

**Comment by hmottestad** [comment]
@rdf4j-bot run tests

**Comment by jeenbroekstra** [comment]
Thanks for this @uditsharma7! The build currently fails with a test failure in the SpinSail - from a glance it looks as if it is the test itself that is the issue here but it will need some eyeballs. I would also like to run this through a full compliance test run before merging as there may be side effects in the query planner that won't show up until we go through full end-to-end testing.

File: core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java (hunk 1 above)
Comment (uditsharma7) [reply core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java]: @hmottestad Updated the code. Also replaced Object.equals with Objects.equals in equals() implementation.

**Comment by uditsharma7** [comment]
@jeenbroekstra I think [Test5](https://github.com/eclipse/rdf4j/blob/master/core/spin/src/test/resources/testcases/test5.ttl) is correct. It is also present [here](https://spinrdf.org/sp.html#sp-Values). 
In the [Unit Test](https://github.com/eclipse/rdf4j/blob/master/core/spin/src/test/java/org/eclipse/rdf4j/spin/SpinParserTest.java#L68) the text query is parsed as 
```
Projection
   ProjectionElemList
      ProjectionElem "country"
      ProjectionElem "code"
   Extension
      ExtensionElem (country)
         Var (name=country)
      ExtensionElem (code)
         Var (name=code)
      BindingSetAssignment ([[country=ex:Germany;code="de"], [country=ex:Norway;code="no"], [country=ex:Endor]])
```
and the spin query is parsed as 
```
Projection
   ProjectionElemList
      ProjectionElem "country"
      ProjectionElem "code"
   Extension
      ExtensionElem (country)
         Var (name=country)
      ExtensionElem (code)
         Var (name=code)
      BindingSetAssignment ([[country=ex:Germany;code="de"], [country=ex:Norway;code="no"], [country=ex:Endor;code=http://spinrdf.org/sp#undef]])
```
It seems there is another bug in [SpinParser.java](https://github.com/eclipse/rdf4j/blob/master/core/spin/src/main/java/org/eclipse/rdf4j/spin/SpinParser.java#L1383) due to which it is treating `sp:undef` as an IRI(http://spinrdf.org/sp#undef) while it should be treated as missing binding.

File: core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java [hunk 2]
```diff
@@ -49,13 +50,19 @@
 	}
 
 	@Override
-	public boolean equals(Object other) {
-		return other instanceof BindingSetAssignment;
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null || getClass() != obj.getClass())
```
Comment (hmottestad) [inline core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java]: Ref: https://stackoverflow.com/questions/1569688/an-efficient-equalsobject-o-implementation

Choice of .getClass() and instanceOf is relevant if we want to allow for subclassing or not. @jeenbroekstra any preference?

Comment (jeenbroekstra) [reply core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/BindingSetAssignment.java]: I tend to prefer allowing subclasses, so I'd use `!(obj instanceof BindingSetAssignment)` instead.

**Comment by jeenbroekstra** [comment]
> It seems there is another bug in [SpinParser.java](https://github.com/eclipse/rdf4j/blob/master/core/spin/src/main/java/org/eclipse/rdf4j/spin/SpinParser.java#L1383) due to which it is treating `sp:undef` as an IRI(http://spinrdf.org/sp#undef) while it should be treated as missing binding.

You're quite right @uditsharma7 . I am running some local tests on your issue branch and will push a fix for this directly to your branch (it's trivial enough that I don't want to bother with a separate bug report).

**Comment by jeenbroekstra** [comment]
add to whitelist

**Pull request merged**